Rewrite the internals of mod configuration remapping to fix bugs

Juuxel opened the pull request:
Fixes #801. Fixes #572.

Needs a bit of testing, esp. with split builds and split dependencies.

- Loom now remaps mods to "collector configs" that are specific to the execution type (compile-time/API vs runtime).
  - For example, `modImplementation` remaps to `modCompileClasspathMain` and `modRuntimeClasspathMain`.
  - This lets us use Gradle's `org.gradle.usage` attributes to select whether it's a compile-time or runtime dependency
  - Note: Remap configurations sourcing from `api` are partially left intact due to their special logic in the `namedElements` configuration.
- Setting the proper usages fixes #801.
- No dependency files are added to the real target configurations (like `api` and `implementation`) anymore.
  - This fixes #572, since `runtimeClasspath` and `compileClasspath` don't leak transitive dependencies unlike `implementation` etc.

Juuxel committed:
Add test for #801

Juuxel committed:
Add test for #572

Juuxel committed:
Rewrite mod configuration remapping internals to fix bugs

Fixes #801. Fixes #572.

- Instead of individual mod configs getting remapped, Loom now remaps
  "collector configs": `mod[Compile/Runtime]Classpath[source set name]`
  (eg `modRuntimeClasspathMain` -> `modRuntimeClasspathMainRemapped`)
  - This lets us use Gradle's `org.gradle.usage` attributes to select
    whether it's a compile-time or runtime dependency
  - Note: Remap configurations sourcing from `api` are partially left
    intact due to their special logic in the `namedElements` configuration.
- Setting the proper usages fixes #801.
- No dependency files are added to the real target configurations
  (like `api` and `implementation`) anymore.
  - This fixes #572, since `runtimeClasspath` and `compileClasspath` don't
    leak transitive dependencies unlike `implementation` etc.

Juuxel committed:
Fix checkstyle

Juuxel commented on src/main/java/net/fabricmc/loom/util/Strings.java:
```diff
@@ -0,0 +1,47 @@
+package net.fabricmc.loom.util;
+
+public final class Strings {
+	public static String capitalize(String word) {
```
I couldn't find this in any libraries and didn't want to pull in another one, let me know if there is a library that we already have that has it

modmuss50 replied:
I think this is fine, I dont see any existing library usage. You could have possibly used SourceSet.getTaskName I think but its clearly not really intended for this usecase.

Juuxel replied:
I used it originally but I need the `main` source set name in this case, otherwise it'd conflict with the generic `modCompileClasspath` used for all compile cp mods.

Juuxel committed:
Fix split env dependency consumers

Juuxel committed:
Only use collector configs for remapped deps and let the inputs stay intact

This way the code has less duplication.

Juuxel committed:
Improve log messages

Juuxel committed:
Update year

Juuxel committed:
Add some comments

Juuxel committed:
Merge remote-tracking branch 'upstream/exp/1.1' into fix/fabric-801

Juuxel committed:
Fix compilation

modmuss50 committed:
Merge remote-tracking branch 'origin/exp/1.1' into fix/fabric-801

Juuxel committed:
Use LinkedHashMap for reliable iteration order through remapped configs

Juuxel committed:
Use ImmutableMap.of instead of Map.of for reliable ordering

Juuxel committed:
ModConfigurationRemapper: Move namedElements handling out of forEach

Juuxel committed:
Add regression test for a crash where loader is resolved after other mods

Juuxel committed:
Fix the aforementioned bug

Juuxel committed:
Rename InstallerDataTest -> DependencyOrderTest

Juuxel committed:
Add TODO about refresh dependencies

The code currently processes the same deps multiple times when
--refresh-dependencies/-Dloom.refresh=true/a cache invalidation
is applied.

modmuss50 reviewed (approved):
Looks great, tests are also useful 👍
